reposerver/repository/repository.go

```diff
@@ -1253,6 +1349,298 @@ func getRepoCredential(repoCredentials []*v1alpha1.RepoCreds, repoURL string) *v
+		"dstPath": dstPath,
+		"hardfail": hardfail,
+	}).Debugf("copyFilesystemSubtree called")
+
+	// Absolute paths are disallowed:
+	if filepath.IsAbs(srcPath) {
+		log.WithFields(log.Fields{
+			"srcRoot": srcRoot,
+			"srcPath": srcPath,
+			"dstRoot": dstRoot,
+			"dstPath": dstPath,
+			"hardfail": hardfail,
+		}).Errorf("copyFilesystemSubtree: srcPath is absolute")
+		return fmt.Errorf("Absolute paths for copy are not allowed: %s", srcPath)
+	}
+	if filepath.IsAbs(dstPath) {
+		log.WithFields(log.Fields{
+			"srcRoot": srcRoot,
+			"srcPath": srcPath,
+			"dstRoot": dstRoot,
+			"dstPath": dstPath,
+			"hardfail": hardfail,
+		}).Errorf("copyFilesystemSubtree: dstPath is absolute")
+		return fmt.Errorf("Absolute paths for copy are not allowed: %s", dstPath)
+	}
+
+	// srcPath and dstPath has to be checked to be within bound of srcRoot and dstRoot
+
+	// First check whether the source file exists
+	srcFullPath := filepath.Join(srcRoot, srcPath)
+	srcStat, err := os.Lstat(srcFullPath)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"srcFullPath": srcFullPath,
+			"srcStat": srcStat,
+		}).Debugf("copyFilesystemSubtree: lstat failed: %v", err)
+		return err
+	}
+
+	// if the source is a symlink, we bail out, because it will certainly be out of bound
+	if srcStat.Mode()&fs.ModeSymlink >0 {
+		log.WithFields(log.Fields{
+			"srcFullPath": srcFullPath,
+			"srcStat": srcStat,
+		}).Debugf("copyFilesystemSubtree: lstat failed: %v", err)
+		return fmt.Errorf("Copy source is a symlink: %s", srcFullPath)
+	}
+
+	// Check whether the source path is within the source root
+	if err := argopath.CheckPathOutOfBound(srcRoot, "", srcPath); err != nil {
+		log.WithFields(log.Fields{
+			"srcRoot": srcRoot,
+			"srcPath": srcFullPath,
+			"srcFullPath": srcFullPath,
+		}).Debugf("copyFilesystemSubtree: argopath.CheckPathOutOfBound failed: %v", err)
+		return err
+	}
+
+	// if we are in hardfail mode, check for any out of bound symlinks
+	if hardfail {
+		if err := argopath.CheckOutOfBoundsSymlinks(srcFullPath); err != nil {
+			log.WithFields(log.Fields{
+				"srcFullPath": srcFullPath,
+			}).Debugf("copyFilesystemSubtree: srcFullPath contains OOB symlinks: %v", err)
+			return err
+		}
+	}
+
+	// Checking whether the destination exists
+	dstFullPath := filepath.Join(dstRoot, dstPath)
+	dstStat, err := os.Lstat(dstFullPath)
+	dstExists := false
+	// check destination to be within bounds:
+	if err == nil {
+		dstExists = true
+		// if the target exists check it
+		if err := argopath.CheckPathOutOfBound(dstRoot, "", dstPath); err != nil {
+			log.WithFields(log.Fields{
+				"dstRoot": dstRoot,
+				"dstPath": dstPath,
+				"dstFullPath": dstFullPath,
+			}).Debugf("copyFilesystemSubtree: argopath.CheckPathOutOfBound failed: %v", err)
+			return err
+		}
+	} else if errors.Is(err, os.ErrNotExist) {
+		// if the destination doesn't yet exist, check its parent
+		if err := argopath.CheckPathOutOfBound(dstRoot, "", filepath.Dir(dstPath)); err != nil {
```

Comment:
What happens if the parent doesn't exist either? For example, the destination path has value `a/b/c/d` and the only directory structure present is `a`?

Reply:
That's quite interesting, let me test it first in the unit tests. What I would like to see, how the current setup behaves, if we have `a`, and specify a dstpath as `a/b/../../../c'. And also how it behaves once creating a path if it doesn't exists. It should fail with being out of bound.

Reply:
So, with the following test:
```
    {testName: "OOB-directory",
      srcRoot: "testdata2/",
      srcPath: "copyfrom-ok",
      dstPath: "dst-dir/b/../../../c",
      hardfail: false,
      shouldFail: false,
    },
```

The test does fail:
```
repository_test.go:2850: [OOB-directory] copyFileSystemSubtree failed: out of bounds symlink found
```
The message is misleading due to reusing the code for OOB symlinks, and this is a directory.

I think it is safe to create the dst directory, but only after checking for OOB.

Reply:
The reason why the destination directory is not created at this part of the code is, if the source is a file and not a directory, then the final part of the path will be a file, and Mkdirall would create that as a directory, and it's not possible to have the same name for both a file and a directory at the same level. That's why the destination directory is created later at the if-else block, specific to what the source it.

Also, it's created with `MkdirAll` in both cases, meaning if multiple level of the destination doesn't exist, then all the missing parts are created.

I've added another unit test for this specific case:
```
    {testName: "OK-multilevel",
      srcRoot: "testdata2/",
      srcPath: "copyfrom-ok",
      dstPath: "dst-dir/a/b/c",
      hardfail: false,
      shouldFail: false,
    },
```

And it's passing:
```
INFO[0000] CopyFrom directory created                    dstFullPath=/home/gczuczy/go/src/github.com/argoproj/argo-cd/reposerver/repository/testdata2/copyfrom-test-3093394955/dst-dir/a/b/c relpath=. targetPath=/home/gczuczy/go/src/github.com/argoproj/argo-cd/reposerver/repository/testdata2/copyfrom-test-3093394955/dst-dir/a/b/c
(...)
PASS
```